File: packages/shared/session-chat-presentation/diff.ts
// Diff extraction for tool calls/results (upstream chat spec §11.4 port).

export const SESSION_CHAT_EDIT_TOOL_NAMES: ReadonlySet<string> = new Set([
  'Edit',
  'MultiEdit',
  'Write',
  'str_replace',
  'apply_patch',
]);

export const SESSION_CHAT_MAX_DIFF_CHARS = 32_000;
export const SESSION_CHAT_DEFAULT_MAX_DIFF_LINES = 120;

export interface SessionChatDiffLine {
  kind: 'add' | 'del' | 'meta' | 'context';
  text: string;
}

const DIFF_TRUNCATED_LINE: SessionChatDiffLine = {
  kind: 'meta',
  text: '… diff truncated …',
};

function toLines(value: unknown, maxLines: number): { lines: string[]; truncated: boolean } {
  if (typeof value !== 'string') {
    return { lines: [], truncated: false };
  }
  const clipped = value.slice(0, SESSION_CHAT_MAX_DIFF_CHARS);
  const lines = clipped.split('\n', maxLines + 1);
  const truncated = value.length > SESSION_CHAT_MAX_DIFF_CHARS || lines.length > maxLines;
  const bounded = lines.slice(0, maxLines);
  if (!truncated && bounded.at(-1) === '') {
    bounded.pop();
  }
  return { lines: bounded, truncated };
}

export function diffFromSessionChatToolCall(
  name: string,
  input: unknown,
  maxLines: number = SESSION_CHAT_DEFAULT_MAX_DIFF_LINES
): SessionChatDiffLine[] | null {
  if (!SESSION_CHAT_EDIT_TOOL_NAMES.has(name) || typeof input !== 'object' || input === null) {
    return null;
  }
  const record = input as Record<string, unknown>;
  const oldValue = record.old_string ?? record.oldString ?? record.old;
  const newValue = record.new_string ?? record.newString ?? record.new ?? record.content ?? record.file_text;
  const oldLines = toLines(oldValue, maxLines);
  const newLines = toLines(newValue, maxLines);
  const deleted: SessionChatDiffLine[] = oldLines.lines.map((text) => ({
    kind: 'del',
    text,
  }));
  const added: SessionChatDiffLine[] = newLines.lines.map((text) => ({
    kind: 'add',
    text,
  }));
  if (deleted.length === 0 && added.length === 0) {
    return null;
  }
  const path = record.file_path ?? record.path;
  const prefix: SessionChatDiffLine[] = typeof path === 'string' ? [{ kind: 'meta', text: path }] : [];
  // ALL dels first, then ALL adds.
  const combined = [...prefix, ...deleted, ...added];
  const truncated = oldLines.truncated || newLines.truncated || combined.length > maxLines;
  return truncated ? [...combined.slice(0, maxLines - 1), DIFF_TRUNCATED_LINE] : combined;
}

/**
 * CDXC:SessionChat 2026-09-18 WHY:
 * Counting leading plus/minus lines mistook Markdown bullets in command output for file changes.
 * Require unified hunk headers and honor their line counts so surrounding output stays uncolored.
 */
export function diffFromSessionChatText(
  text: string,
  maxLines: number = SESSION_CHAT_DEFAULT_MAX_DIFF_LINES
): SessionChatDiffLine[] | null {
  if (text.length === 0) {
    return null;
  }
  const bounded = toLines(text, maxLines);
  let oldRemaining = 0;
  let newRemaining = 0;
  let hasChanges = false;
  const lines: SessionChatDiffLine[] = bounded.lines.map((line) => {
    const hunk = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(?: .*)?\r?$/.exec(line);
    if (hunk) {
      const oldCount = Number(hunk[2] ?? 1);
      const newCount = Number(hunk[4] ?? 1);
      if (!Number.isSafeInteger(oldCount) || !Number.isSafeInteger(newCount)) {
        oldRemaining = newRemaining = 0;
        return { kind: 'context', text: line };
      }
      oldRemaining = oldCount;
      newRemaining = newCount;
      return { kind: 'meta', text: line };
    }
    if (oldRemaining > 0 || newRemaining > 0) {
      if (line.startsWith('+') && newRemaining > 0) {
        newRemaining -= 1;
        hasChanges = true;
        return { kind: 'add', text: line.slice(1) };
      }
      if (line.startsWith('-') && oldRemaining > 0) {
        oldRemaining -= 1;
        hasChanges = true;
        return { kind: 'del', text: line.slice(1) };
      }
      if (line.startsWith(' ') && oldRemaining > 0 && newRemaining > 0) {
        oldRemaining -= 1;
        newRemaining -= 1;
        return { kind: 'context', text: line };
      }
      if (line.replace(/\r$/, '') === '\\ No newline at end of file') {
        return { kind: 'meta', text: line };
      }
      oldRemaining = newRemaining = 0;
    }
    if (/^(?:diff --git |index |--- |\+\+\+ )/.test(line)) {
      return { kind: 'meta', text: line };
    }
    return { kind: 'context', text: line };
  });
  if (!hasChanges) {
    return null;
  }
  return bounded.truncated ? [...lines.slice(0, maxLines - 1), DIFF_TRUNCATED_LINE] : lines;
}

File: packages/gx-chat-core/src/transcript/diff.rs
//! Diff lines: the shape a file-change card counts and paints, and the two extractors that read
//! them out of a tool call or a patch.
//!
//! Ported from `packages/shared/session-chat-presentation/diff.ts`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::transcript::jsstr::{utf16_len, utf16_take};

pub const MAX_DIFF_CHARS: usize = 32_000;
pub const DEFAULT_MAX_DIFF_LINES: usize = 120;

/// What one line of a diff is.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DiffKind {
    Add,
    Del,
    Meta,
    Context,
}

/// One line of a diff, as both renderers paint it.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct DiffLine {
    pub kind: DiffKind,
    pub text: String,
}

impl DiffLine {
    pub fn new(kind: DiffKind, text: impl Into<String>) -> Self {
        Self {
            kind,
            text: text.into(),
        }
    }
}

fn truncated_line() -> DiffLine {
    DiffLine::new(DiffKind::Meta, "\u{2026} diff truncated \u{2026}")
}

/// `value.slice(0, MAX_DIFF_CHARS).split('\n', maxLines + 1)` plus the truncation test.
fn to_lines(value: Option<&str>, max_lines: usize) -> (Vec<String>, bool) {
    let Some(value) = value else {
        return (Vec::new(), false);
    };
    let clipped = utf16_take(value, MAX_DIFF_CHARS);
    // `String.prototype.split` with a limit stops filling the array; it does not join the rest.
    let lines: Vec<&str> = clipped.split('\n').take(max_lines + 1).collect();
    let truncated = utf16_len(value) > MAX_DIFF_CHARS || lines.len() > max_lines;
    let mut bounded: Vec<String> = lines
        .into_iter()
        .take(max_lines)
        .map(str::to_string)
        .collect();
    if !truncated && bounded.last().is_some_and(String::is_empty) {
        bounded.pop();
    }
    (bounded, truncated)
}

fn string_field<'a>(record: &'a Value, keys: &[&str]) -> Option<&'a str> {
    let entries = record.as_object()?;
    for key in keys {
        match entries.get(*key) {
            Some(Value::Null) | None => continue,
            Some(value) => return value.as_str(),
        }
    }
    None
}

/// The tool names whose arguments already hold the before and after of an edit.
pub fn is_edit_tool_name(name: &str) -> bool {
    matches!(
        name,
        "Edit" | "MultiEdit" | "Write" | "str_replace" | "apply_patch"
    )
}

/// The diff a tool call's own arguments describe, before any result comes back.
pub fn diff_from_tool_call(name: &str, input: &Value, max_lines: usize) -> Option<Vec<DiffLine>> {
    if !is_edit_tool_name(name) || !input.is_object() {
        return None;
    }
    let old_value = string_field(input, &["old_string", "oldString", "old"]);
    let new_value = string_field(
        input,
        &["new_string", "newString", "new", "content", "file_text"],
    );
    let (old_lines, old_truncated) = to_lines(old_value, max_lines);
    let (new_lines, new_truncated) = to_lines(new_value, max_lines);
    if old_lines.is_empty() && new_lines.is_empty() {
        return None;
    }
    let path = input
        .get("file_path")
        .or_else(|| input.get("path"))
        .and_then(Value::as_str);
    // ALL dels first, then ALL adds.
    let mut combined: Vec<DiffLine> = path
        .map(|path| DiffLine::new(DiffKind::Meta, path))
        .into_iter()
        .collect();
    combined.extend(
        old_lines
            .into_iter()
            .map(|text| DiffLine::new(DiffKind::Del, text)),
    );
    combined.extend(
        new_lines
            .into_iter()
            .map(|text| DiffLine::new(DiffKind::Add, text)),
    );
    let truncated = old_truncated || new_truncated || combined.len() > max_lines;
    if truncated {
        combined.truncate(max_lines - 1);
        combined.push(truncated_line());
    }
    Some(combined)
}

/// `^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(?: .*)?\r?$`.
fn parse_hunk_header(line: &str) -> Option<(u64, u64)> {
    let line = line.strip_suffix('\r').unwrap_or(line);
    let rest = line.strip_prefix("@@ -")?;
    let (old_count, rest) = parse_hunk_range(rest)?;
    let rest = rest.strip_prefix(" +")?;
    let (new_count, rest) = parse_hunk_range(rest)?;
    let rest = rest.strip_prefix(" @@")?;
    if !rest.is_empty() && !rest.starts_with(' ') {
        return None;
    }
    Some((old_count, new_count))
}

fn parse_hunk_range(value: &str) -> Option<(u64, &str)> {
    let start = value.bytes().take_while(u8::is_ascii_digit).count();
    if start == 0 {
        return None;
    }
    let rest = &value[start..];
    let Some(after_comma) = rest.strip_prefix(',') else {
        return Some((1, rest));
    };
    let digits = after_comma.bytes().take_while(u8::is_ascii_digit).count();
    if digits == 0 {
        return None;
    }
    Some((after_comma[..digits].parse().ok()?, &after_comma[digits..]))
}

/// CDXC:SessionChat 2026-09-18 WHY:
/// Counting leading plus/minus lines mistook Markdown bullets in command output for file changes.
/// Require unified hunk headers and honor their line counts so surrounding output stays uncolored.
pub fn diff_from_text(text: &str, max_lines: usize) -> Option<Vec<DiffLine>> {
    if text.is_empty() {
        return None;
    }
    let (bounded, truncated) = to_lines(Some(text), max_lines);
    let mut old_remaining = 0u64;
    let mut new_remaining = 0u64;
    let mut has_changes = false;
    let mut lines: Vec<DiffLine> = Vec::with_capacity(bounded.len());
    for line in bounded {
        if let Some((old_count, new_count)) = parse_hunk_header(&line) {
            old_remaining = old_count;
            new_remaining = new_count;
            lines.push(DiffLine::new(DiffKind::Meta, line));
            continue;
        }
        if old_remaining > 0 || new_remaining > 0 {
            if line.starts_with('+') && new_remaining > 0 {
                new_remaining -= 1;
                has_changes = true;
                lines.push(DiffLine::new(DiffKind::Add, &line[1..]));
                continue;
            }
            if line.starts_with('-') && old_remaining > 0 {
                old_remaining -= 1;
                has_changes = true;
                lines.push(DiffLine::new(DiffKind::Del, &line[1..]));
                continue;
            }
            if line.starts_with(' ') && old_remaining > 0 && new_remaining > 0 {
                old_remaining -= 1;
                new_remaining -= 1;
                lines.push(DiffLine::new(DiffKind::Context, line));
                continue;
            }
            if line.strip_suffix('\r').unwrap_or(&line) == "\\ No newline at end of file" {
                lines.push(DiffLine::new(DiffKind::Meta, line));
                continue;
            }
            old_remaining = 0;
            new_remaining = 0;
        }
        let meta = ["diff --git ", "index ", "--- ", "+++ "]
            .iter()
            .any(|prefix| line.starts_with(prefix));
        lines.push(DiffLine::new(
            if meta {
                DiffKind::Meta
            } else {
                DiffKind::Context
            },
            line,
        ));
    }
    if !has_changes {
        return None;
    }
    if truncated {
        lines.truncate(max_lines - 1);
        lines.push(truncated_line());
    }
    Some(lines)
}
